src/lib/SaferNames/Syntax.hs

```diff
@@ -107,43 +100,53 @@ data Expr n =
  | Hof (Hof n)
    deriving (Show)
 
+data AtomBinderInfo (n::S) =
+   LetBound LetAnn (Expr n)
+ | LamBound Arrow
+ | PiBound
+ | MiscBound
+ | InferenceName
+
+data TypedBinderInfo (n::S) = TypedBinderInfo (Type n) (AtomBinderInfo n)
+
 data Decl n l = Let LetAnn (Binder n l) (Expr n)
                 deriving (Show)
 
-data DataConRefBinding n l = DataConRefBinding (Binder n l) (Atom n)
-                             deriving Show
+type AtomName   = Name TypedBinderInfo
 
-type AltP (e::E) = Abs (Nest Binder) e :: E
-type Alt = AltP Block                  :: E
+data Binder (n::S) (l::S) =
+  (:>) (NameBinder TypedBinderInfo n l) (Type n)  deriving Show
 
--- TODO: get rid of annotations on variable occurrences
-data AnnVarP (ann::E) n = AnnVar (Name n) (ann n)
-                          deriving Show
+data DataConRefBinding n l = DataConRefBinding (Binder n l) (Atom n) deriving Show
 
-type Var = AnnVarP Type :: E
-type Binder     = AnnBinderP PlainBinder            Type  :: B
-type BinderList = AnnBinderP PlainBinderList (ListE Type) :: B
+type AltP (e::E) = Abs (Nest Binder) e :: E
+type Alt = AltP Block                  :: E
 
 data DataDef n where
-  -- The `RawName` is just for pretty-printing. The actual alpha-renamable binder name is in
-  -- UExpr and Bindings
-  DataDef :: RawName -> BinderList n l -> ListE DataConDef l -> DataDef n
+  -- The `RawName` is just for pretty-printing. The actual alpha-renamable
+  -- binder name is in UExpr and Bindings
+  DataDef :: RawName -> Nest Binder n l -> [DataConDef l] -> DataDef n
 
-data DataConDef n = DataConDef RawName (Abs (Nest Binder) UnitE n)
+-- As above, the `RawName` is just for pretty-printing
+data DataConDef n = DataConDef RawName (EmptyAbs (Nest Binder) n)
                     deriving Show
 
-data NamedDataDef n = NamedDataDef (Name n) (DataDef n)
-                      deriving Show
+data Block n where
+  Block :: Type n -> Nest Decl n l ->  Expr l -> Block n
```

Comment:
What's this type? Is it the return type? Why can't we query the result expr for that anymore?

Reply:
Yes, it's the return type. The expr lives in the name space under the decls, so querying its type can leak names! We could use `fromConstAbs` to check for leaks, but that then requires traversing the block. This is just a cached result as an optimization (and also a sort of proof that there are no leaks). I'll write a note.